### ISSUE TITLE
how to use alphas

### ISSUE THREAD

**Issue by yoopaan** [SamuelGong/FLASHE#1]
您好，关于https://github.com/SamuelGong/FLASHE/blob/master/encrypt_test/final_big_table.ipynb 中的代码有一些问题请教。

对梯度quantize时会返回一个值"

```
a_q, alphas = quantize([[a]], element_bits)
```

求和之后unquantize时还需要把alphas作为参数传入，
```
a_u = unquantize(a_d, alphas, element_bits, num_clients)
```

这个alphas的作用是什么？

**Comment by SamuelGong** [comment]
Thanks for reaching out! In order to understand the quantization and de-quantization process, I strongly recommend you to have a look at the related source code:

https://github.com/SamuelGong/FLASHE/blob/e69778c5c7ec8c5741dd9076fd33e54a0d5d5ed6/federatedml/secureprotol/jzf_quantize.py#L33

and 

https://github.com/SamuelGong/FLASHE/blob/e69778c5c7ec8c5741dd9076fd33e54a0d5d5ed6/federatedml/secureprotol/jzf_quantize.py#L84

You can even play with them using some toy examples.

Basically, $\alpha$ is used to clip the initial values during quantization, so that they are located in the range $[-\alpha, \alpha]$ before being transferred to integers. Thus, during de-quantization, the program needs to map the quantized integers back to floating-point numbers in the range $[-\alpha, \alpha]$.

**Comment by yoopaan** [comment]
Thanks a lot for your prompt reply！
I made a simple try，but the result seems not right.

Can you point out what is wrong with my code? Thx.
```
def phe_single_enc(a):
    a_q, alphas = quantize([[a]], element_bits)
    a_q = a_q[0]
    a_e = paillier_encrypt(a_q)
    return a_e, alphas

def phe_single_dec(a_e, alphas):
    dec_a = paillier_decrypt(a_e)
    dec_a = unquantize(dec_a, alphas, element_bits, num_clients)
    return dec_a


def phe_batch_enc(a):
    a_q, alphas = quantize([[a]], element_bits)
    a_q = a_q[0]
    shape = a_q.shape
    a_q = _static_batching_padding(a_q, paillier.key_length, element_bits, additional_bits)
    a_e = paillier_encrypt(a_q)
    return a_e, shape, alphas


def phe_batch_dec(a_e, shape, alphas):
    a_d = paillier_decrypt(a_e)
    a_d = _static_unbatching_padding(a_d, paillier.key_length, element_bits, additional_bits)
    a_d = a_d[:(int(np.prod(shape)))]
    a_d = a_d.reshape(shape)
    a_d = unquantize(a_d, alphas, element_bits, num_clients)
    return a_d


def enc_sum(enc_list):
    mod = paillier.get_n() ** 2
    enc_add_result = reduce(lambda x, y: (x * y) % mod, enc_list)
    return enc_add_result

arr_size = 10
a1 = np.random.random(arr_size)
a2 = np.random.random(arr_size)
a3 = np.random.random(arr_size)


enc_a1, alphas1 = phe_single_enc(a1)
enc_a2, alphas2 = phe_single_enc(a2)
enc_a3, alphas3 = phe_single_enc(a3)

enc_a1a2a3 = enc_sum([enc_a1,enc_a2,enc_a3])


phe_single_dec(enc_a1a2a3, [alphas1,alphas2,alphas3])
```
My result:
![1](https://user-images.githubusercontent.com/38783332/194907439-c359c3d0-f521-4880-9696-dd3852084a86.png)

**Comment by SamuelGong** [comment]
In your example, the expected behavior seems to be that Out[99] matches Out[103], if my understandings are correct.

Now for the result, you have obtained, it may or may not be the expected one. If you have not clipped too much information, i.e., almost all your values falling into the range $[\alpha, -\alpha]$, the two results should be very close. On the other hand, if you are using a relatively small $\alpha$, i.e., some of your values got cut during the clipping, chances are that the two results do not match exactly.

In other words, clipping does cause a loss of information, so currently, I cannot tell whether there is something wrong or it is just what you should have for a correct program.

**Comment by yoopaan** [comment]
The complete test code:
```
import numpy as np
import time
from multiprocessing import Pool, cpu_count
from functools import reduce
import compress_pickle

element_bits = 16
num_clients = 3
MAGIC_N_JOBS = 50
additional_bits = int(np.ceil(np.log2(num_clients + 1)))
int_bits = element_bits + additional_bits

### Quantization
from federatedml.secureprotol.jzf_aciq import ACIQ
from federatedml.secureprotol.jzf_quantize import     _static_quantize_padding, _static_unquantize_padding



def get_alpha_r_max(trainable_list, element_bits):
    local_min_list = []
    local_max_list = []
    size_list = []
    for idx, trainable in enumerate(trainable_list):
        local_min = []
        local_max = []
        for layer in trainable:
            local_min.append(np.amin(layer))
            local_max.append(np.amax(layer))

            if idx == 0:
                size_list.append(layer.size)

        local_min_list.append(np.array(local_min))
        local_max_list.append(np.array(local_max))

    local_min_list = np.array(local_min_list)
    local_max_list = np.array(local_max_list)

    min_list = np.amin(local_min_list, 0)
    max_list = np.amax(local_max_list, 0)

    n = len(trainable_list)
    aciq = ACIQ(element_bits)
    alpha_list = []
    r_max_list = []
    layer_cnt = 0
    for min, max in zip(min_list, max_list):
        alpha = aciq.get_alpha_gaus(min, max, size_list[layer_cnt])

        r_max = alpha * num_clients

        alpha_list.append(alpha)
        r_max_list.append(r_max)
        layer_cnt += 1

    return alpha_list, r_max_list


def quantize(trainable_list, element_bits):
    n = len(trainable_list)
    alpha_list, r_max_list = get_alpha_r_max(trainable_list,
                                            element_bits)
    quantized = []
    for _, trainable in enumerate(trainable_list):
        quantized_layers = []
        for idx, layer in enumerate(trainable):
            shape = layer.shape
            layer_flatten = layer.flatten()
            ret = _static_quantize_padding(layer_flatten,
                                           alpha_list[idx],
                                           element_bits,
                                           n)
            ret = np.array(ret).reshape(shape)
            quantized_layers.append(ret)
        quantized.append(np.array(quantized_layers))
    return np.array(quantized), np.array(alpha_list)

def unquantize(trainable, alpha_list, element_bits, num_clients):
    layers = []
    for idx, layer in enumerate(trainable):
        shape = layer.shape
        layer_flatten = layer.flatten()
        ret = _static_unquantize_padding(layer_flatten,
                                         alpha_list[idx],
                                         element_bits,
                                         num_clients)
        ret = np.array(ret).reshape(shape)
        layers.append(ret)

    layers = np.array(layers)
    return layers

from federatedml.secureprotol.jzf_quantize import _static_batching_padding, _static_unbatching_padding


def chunks_idx(l, n):
    d, r = divmod(len(l), n)
    for i in range(n):
        si = (d+1)*(i if i < r else r) + d*(0 if i < r else i - r)
        yield si, si+(d+1 if i < r else d)\
def _compress(flatten_array, num_bits):
    res = 0
    l = len(flatten_array)
    for element in flatten_array:
        res <<= num_bits
        res += element
    return res, l


def compress_multi(flatten_array, num_bits):
    l = len(flatten_array)
    pool_inputs = []
    sizes = []
    pool = Pool(MAGIC_N_JOBS)
    for begin, end in chunks_idx(range(l), MAGIC_N_JOBS):
        sizes.append(end - begin)
        pool_inputs.append([flatten_array[begin:end], num_bits])
    pool_outputs = pool.starmap(_compress, pool_inputs)
    pool.close()
    pool.join()
    res = 0
    for idx, output in enumerate(pool_outputs):
        res +=  output[0] << (int(np.sum(sizes[idx + 1:])) * num_bits)
    
    num_bytes = (num_bits * l - 1) // 8 + 1
    res = res.to_bytes(num_bytes, 'big')
    return res, l


from federatedml.secureprotol.jzf_paillier import PaillierCipher
paillier = PaillierCipher()
paillier.generate_key(n_length=1024)

def paillier_encrypt(value):
    return paillier.encrypt(value)

def paillier_decrypt(value):
    return paillier.decrypt(value)


# In[114]:


def phe_single_enc(a):
    a_q, alphas = quantize([[a]], element_bits)
    a_q = a_q[0]
    a_e = paillier_encrypt(a_q)
    return a_e, alphas

def phe_single_dec(a_e, alphas):
    dec_a = paillier_decrypt(a_e)
    dec_a = unquantize(dec_a, alphas, element_bits, num_clients)
    return dec_a


def phe_batch_enc(a):
    a_q, alphas = quantize([[a]], element_bits)
    a_q = a_q[0]
    shape = a_q.shape
    a_q = _static_batching_padding(a_q, paillier.key_length, element_bits, additional_bits)
    a_e = paillier_encrypt(a_q)
    return a_e, shape, alphas


def phe_batch_dec(a_e, shape, alphas):
    a_d = paillier_decrypt(a_e)
    a_d = _static_unbatching_padding(a_d, paillier.key_length, element_bits, additional_bits)
    a_d = a_d[:(int(np.prod(shape)))]
    a_d = a_d.reshape(shape)
    a_d = unquantize(a_d, alphas, element_bits, num_clients)
    return a_d


def enc_sum(enc_list):
    mod = paillier.get_n() ** 2
    enc_add_result = reduce(lambda x, y: (x * y) % mod, enc_list)
    return enc_add_result




arr_size = 10





a1 = np.random.random(arr_size)
a2 = np.random.random(arr_size)
a3 = np.random.random(arr_size)


a1+a2+a3



enc_a1, alphas1 = phe_single_enc(a1)
enc_a2, alphas2 = phe_single_enc(a2)
enc_a3, alphas3 = phe_single_enc(a3)




[alphas1,alphas2,alphas3]



enc_a1a2a3 = enc_sum([enc_a1,enc_a2,enc_a3])





phe_single_dec(enc_a1a2a3, [alphas1,alphas2,alphas3])


```

**Comment by SamuelGong** [comment]
It seems to me a better idea to going to the definition of `_static_quantize_padding` function (I have pasted the link above), and doing something like:

```Python
def _static_quantize_padding(value, alpha, int_bits, num_clients):
    # first clipping
    print(f'Before: {value[:10]}')
    value = np.clip(value, -alpha, alpha)
    print(f'After: {value[:10]}')

   ...
````

Then you should see whether clipping caused loss of information. If it did, you may consider your obtained results as normal ones.

**Comment by yoopaan** [comment]
> 

Many thanks!

**Comment by SamuelGong** [comment]
Glad that I can help you! Please consider citing our work if it is helpful in developing your work and you happen to write related academic papers. Thanks!
